src/models/llama/decoder.rs

```diff
@@ -0,0 +1,279 @@
+            name = name.replace("attention.output", "attention.o_proj");
+            name = name.replace("attention.layer_norm", "input_layernorm");
+            name = name.replace("attention.", "self_attn.");
+
+            // Feed-forward layer.
+            name = name.replace("ffn.layer_norm", "post_attention_layernorm");
+            name = name.replace("ffn.output", "ffn.down_proj");
+            name = name.replace("ffn.", "mlp.");
+            name = name.replace("intermediate", "up_proj");
+            name = name.replace("gate", "gate_proj");
+
+            // Layer norm after all layers.
+            name = name.replace("output_layer_norm", "norm");
+
+            // Output vocab.
+            name = name.replace("output_embeddings", "lm_head");
+
+            static LAYER_RE: OnceLock<Regex> = OnceLock::new();
+            let layer_re =
+                LAYER_RE.get_or_init(|| Regex::new(r"layer_(\d+)").expect("Invalid regex"));
+            name = layer_re.replace(&name, "layers.$1").to_string();
+            name
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use candle_core::{DType, Device, Tensor};
+    use ndarray::array;
+    use snafu::{report, FromString, ResultExt, Whatever};
+
+    use crate::architectures::{Decoder, LayerOutputs};
+    use crate::kv_cache::KeyValueCache;
+    use crate::layers::attention::AttentionMask;
+    use crate::models::hf::FromHFHub;
+    use crate::models::llama::LlamaDecoder;
+    use crate::util::tests::assert_tensor_eq;
+
+    fn sample_inputs() -> Result<(Tensor, Tensor), Whatever> {
+        let input = Tensor::arange(0i64, 24, &Device::Cpu)
+            .and_then(|t| t.reshape((3, 8)))
+            .with_whatever_context(|_| "Cannot create input tensor")?;
+
+        let mask = Tensor::from_slice(
+            &[
+                1u32, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0,
+            ],
+            (3, 8),
+            &Device::Cpu,
+        )
+        .with_whatever_context(|_| "Cannot create attention mask tensor")?;
+
+        Ok((input, mask))
+    }
+
+    #[test]
+    #[report]
+    fn llama_decoder_emits_correct_output() -> Result<(), Whatever> {
+        let decoder =
+            LlamaDecoder::from_hf_hub("explosion-testing/llama2-kv-sharing", None, Device::Cpu)
+                .with_whatever_context(|_| "Cannot load model")?;
+
+        let (input, mask) = sample_inputs()?;
+
+        let output = decoder
+            .forward_t(
+                &input,
+                &AttentionMask::new(mask).whatever_context("Cannot build attention mask")?,
+                &mut KeyValueCache::no_cache(5),
+                None,
+                false,
+            )
+            .map_err(|e| Whatever::with_source(e, "Cannot decode input".to_string()))?;
+
+        let summed_states = output
+            .layer_outputs()
+            .last()
+            .unwrap()
+            .sum(2)
```

Comment:
I imagine we're checking the aggregates in these tests so as to not have huge ground truth tensors in the code, but is it sufficient? Won't it also make sense to use an operation that controls for position-variance, i.e., a `gather` op? Same below.

I remember that running into a bug of this nature when working on DeBERTa in `curated-transformers`, where the aggregates were correct but the outputs were incorrect due to their dimensions being swapped.

Reply:
Maybe, we have to find some balance between catching everything, but still having small test vectors. Maybe a small gather with prime-based indexes (to avoid missing mismatches when things are swapped in a regular way).

Reply:
A small `gather` test sounds good.

Reply:
Added the aggregates using dot products with pseudo-random vectors.